### PR TITLE
CIVDT-303: Item Number Page

### PR DESCRIPTION
**Pull request by jakebooker56**

### CODE REVIEW

File: app/controllers/ItemNumberController.scala [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+                                                    view: ItemNumberView,
+                                                    formProvider: ItemNumberFormProvider
+                                                   )
+  extends FrontendController(mcc) with I18nSupport {
+
+  def onLoad: Action[AnyContent] = (identify andThen getData andThen requireData).async { implicit request =>
+    val form = request.userAnswers.get(ItemNumberPage).fold(formProvider()) {
+      formProvider().fill
+    }
+    Future.successful(Ok(view(form,backLink)))
```
Comment (tvaidotas) [inline app/controllers/ItemNumberController.scala:46]:
```suggestion
    Future.successful(Ok(view(form, backLink)))
```

File: app/controllers/ItemNumberController.scala [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  }
+
+  def onSubmit: Action[AnyContent] = (identify andThen getData andThen requireData).async { implicit request =>
+    formProvider().bindFromRequest().fold(
+      formWithErrors => Future.successful(BadRequest(view(formWithErrors,backLink))),
```
Comment (tvaidotas) [inline app/controllers/ItemNumberController.scala:51]:
```suggestion
      formWithErrors => Future.successful(BadRequest(view(formWithErrors, backLink))),
```

File: app/controllers/ItemNumberController.scala [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+          updatedAnswers <- Future.fromTry(request.userAnswers.set(ItemNumberPage, value))
+          _ <- sessionRepository.set(updatedAnswers)
+        } yield {
+          Redirect(controllers.routes.ItemNumberController.onLoad()) // values controller
+        }
+      }
+    )
+  }
+
+  private[controllers] def backLink: Call = Call("GET",controllers.routes.BoxNumberController.onLoad().url)
```
Comment (tvaidotas) [inline app/controllers/ItemNumberController.scala:63]:
```suggestion
  private[controllers] def backLink: Call = Call("GET", controllers.routes.BoxNumberController.onLoad().url)
```

Comment (tvaidotas) [inline app/controllers/ItemNumberController.scala:63]: change it to a private lazy val

File: app/pages/ItemNumberPage.scala [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package pages
+
+import models.CustomsProcedureCode
```
Comment (tvaidotas) [inline app/pages/ItemNumberPage.scala:19]: isnt used

File: app/forms/ItemNumberFormProvider.scala [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+import forms.mappings.Mappings
+import javax.inject.Inject
+import play.api.data.Form
+import play.api.i18n.Messages
+
+
+class ItemNumberFormProvider @Inject() extends Mappings {
+
+  def apply()(implicit messages: Messages): Form[String] =
```
Comment (tvaidotas) [inline app/forms/ItemNumberFormProvider.scala:27]: should be an Int

File: app/pages/ItemNumberPage.scala [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+import play.api.libs.json.JsPath
+
+case object ItemNumberPage extends QuestionPage[String] {
```
Comment (tvaidotas) [inline app/pages/ItemNumberPage.scala:22]: should be an int

File: app/views/ItemNumberView.scala.html [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+   @layout(
+    pageTitle = messages("itemNo.pageTitle"),
+    form = Some(form),
+    customBackLinkUrl = Some(backLink)) {
+
+    @errorSummary(form.errors)
+
+    @formHelper(action = controllers.routes.ItemNumberController.onSubmit()) {
+      @inputText(form, "itemNumber", "itemNumber", messages("itemNo.pageHeader"), true, classes = Some("govuk-fieldset__legend govuk-input--width-10"))
```
Comment (tvaidotas) [inline app/views/ItemNumberView.scala.html:35]: should be inputNumber

File: app/views/ItemNumberView.scala.html [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *@
+
+@this(layout: templates.Layout,
+      formHelper: FormWithCSRF,
+      h1: components.h1,
```
Comment (tvaidotas) [inline app/views/ItemNumberView.scala.html:19]: you;re not using it

File: app/views/ItemNumberView.scala.html [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+      errorSummary: components.errorSummary,
+      button: components.button,
+      inputText: components.inputText
+)
+
+@(form: Form[_],backLink: Call)(implicit request: Request[_], messages: Messages)
```
Comment (tvaidotas) [inline app/views/ItemNumberView.scala.html:25]:
```suggestion
@(form: Form[_], backLink: Call)(implicit request: Request[_], messages: Messages)
```

File: test/controllers/ItemNumberControllerSpec.scala [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+      dataRequiredAction,
+      mockSessionRepository,
+      messagesControllerComponents,
+      ItemNumberView,
+      form
+    )
+    private lazy val ItemNumberView = app.injector.instanceOf[ItemNumberView]
+    private lazy val dataRetrievalAction = new FakeDataRetrievalAction(userAnswers)
+    val userAnswers: Option[UserAnswers] = userAnswersWithItemNumber
+    val formProvider: ItemNumberFormProvider = injector.instanceOf[ItemNumberFormProvider]
```
Comment (tvaidotas) [inline test/controllers/ItemNumberControllerSpec.scala:58]:
```suggestion
    private lazy val dataRetrievalAction = new FakeDataRetrievalAction(userAnswersWithItemNumber)
    val formProvider: ItemNumberFormProvider = injector.instanceOf[ItemNumberFormProvider]
```

File: test/controllers/ItemNumberControllerSpec.scala [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+    }
+
+  }
+
+  "POST /" when {
+
+    "payload contains valid data" should {
+
+      "return a SEE OTHER response when correct data with numeric only values" in new Test {
+        override val userAnswers: Option[UserAnswers] = userAnswersWithItemNumber
```
Comment (tvaidotas) [inline test/controllers/ItemNumberControllerSpec.scala:82]:
```suggestion
```

File: test/controllers/ItemNumberControllerSpec.scala [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+        lazy val result: Future[Result] = controller.onSubmit(fakeRequestGenerator("1"))
+        status(result) mustBe Status.SEE_OTHER
+      }
+
+      "update the UserAnswers in session" in new Test {
+        override val userAnswers: Option[UserAnswers] = userAnswersWithItemNumber
```
Comment (tvaidotas) [inline test/controllers/ItemNumberControllerSpec.scala:88]:
```suggestion
```

File: test/controllers/ItemNumberControllerSpec.scala [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+        await(controller.onSubmit(fakeRequestGenerator("1")))
+        verifyCalls()
+      }
+
+    }
+
+    "payload contains invalid data" should {
+
+      "return BAD REQUEST when invalid data is sent" in new Test {
+        override val userAnswers: Option[UserAnswers] = userAnswersWithItemNumber
```
Comment (tvaidotas) [inline test/controllers/ItemNumberControllerSpec.scala:98]:
```suggestion
```

**Review by tvaidotas — CHANGES_REQUESTED** [review]

File: app/controllers/ItemNumberController.scala (hunk 3 above)
Comment (ian-ralphs) [reply app/controllers/ItemNumberController.scala:63]: and move it to the top of the class

File: app/views/ItemNumberView.scala.html (hunk 7 above)
Comment (ian-ralphs) [reply app/views/ItemNumberView.scala.html:35]: and can you make the width 5 please

File: conf/messages [hunk 14]
```diff
@@ -275,3 +275,13 @@ cya.payingByDeferment = By deferment?
 cya.numberOfAmendments = Number of amendments
 cya.supportingInformation = Supporting information
 cya.newCustomsProcedure = Amended customs procedure code
+
+# Item Number
+itemNo.pageTitle = What is the item number?
+itemNo.pageHeader = What is the item number?
```
Comment (ian-ralphs) [inline conf/messages:281]: Can you use itemNumber throughout please

**Review by ian-ralphs — CHANGES_REQUESTED** [review]

**Pull request merged**